extensions/pkg/webhook/cloudprovider/mutator.go

```diff
@@ -0,0 +1,103 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cloudprovider
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/gardener/gardener/extensions/pkg/webhook"
+	"github.com/gardener/gardener/extensions/pkg/webhook/controlplane/genericmutator"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/inject"
+)
+
+// Ensurer ensures that the cloudprovider secret conforms to the provider requirements.
+type Ensurer interface {
+	EnsureCloudProviderSecret(ctx context.Context, etcx genericmutator.EnsurerContext, new, old *corev1.Secret) error
+}
+
+// NewMutator creates a new cloudprovider mutator.
+func NewMutator(logger logr.Logger, ensurer Ensurer) webhook.Mutator {
+	return &mutator{
+		logger:  logger.WithName("mutator"),
+		ensurer: ensurer,
+	}
+}
+
+type mutator struct {
+	client  client.Client
+	logger  logr.Logger
+	ensurer Ensurer
+}
+
+// InjectClient injects the client into the ensurer.
+func (m *mutator) InjectClient(client client.Client) error {
+	m.client = client
+	if _, err := inject.ClientInto(client, m.ensurer); err != nil {
+		return fmt.Errorf("could not inject the client into the ensurer: %v", err)
+	}
+	return nil
+}
+
+// InjectScheme injects the manager's scheme into the ensurer.
+func (m *mutator) InjectScheme(scheme *runtime.Scheme) error {
+	if _, err := inject.SchemeInto(scheme, m.ensurer); err != nil {
+		return fmt.Errorf("could not inject scheme into the ensurer: %v", err)
+	}
+	return nil
+}
+
+// Mutate validates and if needed mutates the given object.
+func (m *mutator) Mutate(ctx context.Context, new, old runtime.Object) error {
+	acc, err := meta.Accessor(new)
+	if err != nil {
+		return fmt.Errorf("could not create accessor during webhook: %v", err)
+	}
+
+	if acc.GetDeletionTimestamp() != nil {
+		return nil
+	}
+
+	newSecret, ok := new.(*corev1.Secret)
+	if !ok {
+		return fmt.Errorf("could not mutate: object is not of type %q", "Secret")
+	}
+	if newSecret.Name != v1beta1constants.SecretNameCloudProvider {
+		return nil
+	}
+
+	var oldSecret *corev1.Secret
+	if old != nil {
+		oldSecret, ok = old.(*corev1.Secret)
+		if !ok {
+			return fmt.Errorf("could not mutate: old object could not be casted to type %q", "Secret")
+		}
+	}
+
+	o, ok := new.(metav1.Object)
+	if !ok {
+		return fmt.Errorf("could not cast runtime object to %q", "metav1.Object")
+	}
+	etcx := genericmutator.NewEnsurerContext(m.client, o)
```

Comment:
Can you move this `EnsurerContext` out from the `genericmutator` into some other package? It's not specific to the generic controlplane mutator anymore now.

Reply:
Alternatively, maybe duplicate it locally in the package ? They are the same because I need to fetch the `Cluster` but this is a context specific to this admission webhook, so the similarity is "coincidental".

Reply:
Hm, I see what you mean, however, does this really bring any meaningful benefit? Both mutators need this context, and it seems unlikely (as of today, at least) that they will be different in the future. For now, I think duplicating it would not help in anyway. As long as we can share the code we should do it to reduce the maintenance efforts. If it becomes necessary to have dedicated contexts in the future then we are not blocked and can still do it.